File: dotnet/source/examples/Publishing/TimeSeriesAppendWithTimestamp.cs
/**
 * Copyright © 2020 Push Technology Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Threading;
using System.Threading.Tasks;
using PushTechnology.ClientInterface.Client.Callbacks;
using PushTechnology.ClientInterface.Client.Factories;
using PushTechnology.ClientInterface.Client.Features.Control.Topics;
using PushTechnology.ClientInterface.Client.Session;
using PushTechnology.ClientInterface.Client.Topics;
using PushTechnology.ClientInterface.Client.Topics.Details;
using static System.Console;
using static PushTechnology.ClientInterface.Examples.Runner.Program;

namespace PushTechnology.ClientInterface.Example.Publishing {
    /// <summary>
    /// Control client implementation that appends values with a user supplied timestamp to a time series topic.
    /// </summary>
    public sealed class PublishingTimeSeriesAppendWithTimestamp : IExample {
        /// <summary>
        /// Runs the time series topic append (with timestamp) control client example.
        /// </summary>
        /// <param name="cancellationToken">A token used to end the client example.</param>
        /// <param name="args">A single string should be used for the server url.</param>
        public async Task Run( CancellationToken cancellationToken, string[] args ) {
            string TOPIC_PREFIX = "time-series";

            var serverUrl = args[ 0 ];
            var session = Diffusion.Sessions.Principal( "control" ).Password( "password" )
                .CertificateValidation((cert, chain, errors) => CertificateValidationResult.ACCEPT)
                .Open( serverUrl );

            // Create a string topic
            var typeName = Diffusion.DataTypes.Get<String>().TypeName;
            var topicPath = $"{TOPIC_PREFIX}/{typeName}/{DateTime.Now.ToFileTimeUtc()}";
            var specification = session.TopicControl.NewSpecification(TopicType.TIME_SERIES)
                .WithProperty(TopicSpecificationProperty.TimeSeriesEventValueType, typeName);

            try
            {
                await session.TopicControl.AddTopicAsync(topicPath, specification, cancellationToken);
            }
            catch (Exception ex)
            {
                WriteLine($"Failed to add topic '{topicPath}' : {ex}.");
                session.Close();
                return;
            }

            try
            {
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value1", DateTimeOffset.FromUnixTimeMilliseconds(322), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value1", DateTimeOffset.FromUnixTimeMilliseconds(323), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value1", DateTimeOffset.FromUnixTimeMilliseconds(323), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value1", DateTimeOffset.FromUnixTimeMilliseconds(324), cancellationToken);
                await session.TimeSeries.AppendAsync<String>(topicPath, "Value1", DateTimeOffset.FromUnixTimeMilliseconds(325), cancellationToken);

                await Task.Delay(TimeSpan.FromMilliseconds(300));
            }
            catch (Exception ex)
            {
                WriteLine($"Topic {topicPath} value could not be appended : {ex}.");
            }

            // Remove the string topic
            try
            {
                await session.TopicControl.RemoveTopicsAsync( topicPath, cancellationToken );
            } catch(Exception ex) {
                WriteLine( $"Failed to remove topic '{topicPath}' : {ex}." );
            }

            // Close the session
            session.Close();
        }
    }
}

File: python/examples/publishing/time_series_append_with_timestamp.py
#  Copyright (c) 2022 - 2023 DiffusionData Ltd., All Rights Reserved.
#
#  Use is subject to licence terms.
#
#  NOTICE: All information contained herein is, and remains the
#  property of DiffusionData. The intellectual and technical
#  concepts contained herein are proprietary to DiffusionData and
#  may be covered by U.S. and Foreign Patents, patents in process, and
#  are protected by trade secret or copyright law.


import asyncio
import datetime

import diffusion
import diffusion.datatypes


# Diffusion server connection information;
# adjust as needed for the server used in practice.
from diffusion.features.timeseries import TimeSeries

server_url = "ws://localhost:8080"
principal = "admin"
credentials = diffusion.Credentials("password")

TOPIC_PREFIX = "time-series"


async def main():
    """
    Control client implementation that appends values
    with a user supplied timestamp to a time series topic.
    """
    # Creating the session.
    async with diffusion.Session(
        url=server_url, principal=principal, credentials=credentials
    ) as session:
        # Create a string topic
        topic_type = diffusion.datatypes.STRING
        topic_path = f"{TOPIC_PREFIX}/{topic_type.type_name}/{datetime.datetime.utcnow()}"
        specification = TimeSeries.of(topic_type).with_properties()

        try:
            await session.topics.add_topic(topic_path, specification)
        except Exception as ex:
            print(f"Failed to add topic '{topic_path}' : {ex}.")
            await session.close()
            return

        try:
            await session.time_series.append(
                topic_path,
                "Value1",
                diffusion.datatypes.STRING,
                timestamp=datetime.datetime.now()
                + datetime.timedelta(milliseconds=322),
            )
            await session.time_series.append(
                topic_path,
                "Value1",
                diffusion.datatypes.STRING,
                timestamp=datetime.datetime.now()
                + datetime.timedelta(milliseconds=323),
            )
            await session.time_series.append(
                topic_path,
                "Value1",
                diffusion.datatypes.STRING,
                timestamp=datetime.datetime.now()
                + datetime.timedelta(milliseconds=323),
            )
            await session.time_series.append(
                topic_path,
                "Value1",
                diffusion.datatypes.STRING,
                timestamp=datetime.datetime.now()
                + datetime.timedelta(milliseconds=324),
            )
            await session.time_series.append(
                topic_path,
                "Value1",
                diffusion.datatypes.STRING,
                timestamp=datetime.datetime.now()
                + datetime.timedelta(milliseconds=325),
            )

            await asyncio.sleep(0.3)
        except Exception as ex:
            print(f"Topic {topic_path} value could not be appended : {ex}.")

        # Remove the string topic
        try:
            await session.topics.remove_topic(topic_path)
        except Exception as ex:
            print(f"Failed to remove topic '{topic_path}' : {ex}.")
